x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Case.java

```diff
@@ -185,6 +188,43 @@ public Object fold() {
         return elseValue.fold();
     }
 
+    /**
+     * Fold the arms of {@code CASE} statements.
+     * <ol>
+     *     <li>
+     *         Conditions that evaluate to {@code false} are removed so
+     *         {@code EVAL c=CASE(false, foo, b, bar, bort)} becomes
+     *         {@code EVAL c=CASE(b, bar, bort)}.
+     *     </li>
+     *     <li>
+     *         Conditions that evaluate to {@code true} stop evaluation and
+     *         return themselves so {@code EVAL c=CASE(true, foo, bar)} becomes
+     *         {@code EVAL c=foo}.
+     *     </li>
+     * </ol>
+     * And those two combine so {@code EVAL c=CASE(false, foo, b, bar, true, bort, el)} becomes
+     * {@code EVAL c=CASE(b, bar, bort)}.
+     */
+    public Expression partiallyFold() {
+        List<Expression> newChildren = new ArrayList<>(children().size());
+        for (Condition condition : conditions) {
+            if (condition.condition.foldable() == false) {
+                newChildren.add(condition.condition);
+                newChildren.add(condition.value);
+                continue;
+            }
+            Boolean b = (Boolean) condition.condition.fold();
+            if (b != null && b) {
+                newChildren.add(condition.value);
+                return replaceChildren(newChildren);
+            }
+        }
+        if (elseValueIsExplicit()) {
+            newChildren.add(elseValue);
+        }
+        return replaceChildren(newChildren);
```

Comment:
To avoid having to create a new instance, you could add a `boolean modified = false`, set that to true in the call and then change the return to `return modified ? replaceChildren(newChildren) : this)`